### PR TITLE
onCollisionEnd callbacks for Collidable and HitboxShape

### PR DESCRIPTION
**Pull request by spydon**
# Description

Calls `onCollisionEnd` on `Collidable` and `HitboxShape` once a collision is over.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [x] I have formatted my code with `./scripts/format.sh` and the Flame analyzer (`./scripts/analyze.sh`) does not report any problems.
- [x] I read and followed the [Flame Style Guide].
- [x] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [x] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [x] No, this is *not* a breaking change.

## Related Issues

*Replace this paragraph with a list of issues related to this PR from the [issue database]. Indicate, which of these issues are resolved or fixed by this PR.*

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md

### CODE REVIEW

File: examples/lib/stories/collision_detection/multiple_shapes.dart [hunk 1]
```diff
@@ -34,17 +34,18 @@ abstract class MyCollidable extends PositionComponent
     anchor = Anchor.center;
   }
 
+  @override
+  Future<void> onLoad() async {
+    await super.onLoad();
```
Comment (erickzanardo) [inline examples/lib/stories/collision_detection/multiple_shapes.dart:39]: The super onLoad don't really do anything, you can just remove this line.

File: examples/lib/stories/collision_detection/multiple_shapes.dart [hunk 2]
```diff
@@ -71,25 +73,34 @@ abstract class MyCollidable extends PositionComponent
 
   @override
   void onCollision(Set<Vector2> intersectionPoints, Collidable other) {
-    _isHit = true;
+    final isNew = _activeCollisions.add(other);
+    if (isNew) {
+      _activePaint..color = collisionColor(other).withOpacity(0.8);
```
Comment (erickzanardo) [inline examples/lib/stories/collision_detection/multiple_shapes.dart:78]:
```suggestion
      _activePaint.color = collisionColor(other).withOpacity(0.8);
```

File: examples/lib/stories/collision_detection/multiple_shapes.dart [hunk 3]
```diff
@@ -152,20 +163,18 @@ class CollidableCircle extends MyCollidable {
...
   SnowmanPart(double definition, Vector2 relativeOffset, Color hitColor)
       : super(definition: definition) {
     this.relativeOffset.setFrom(relativeOffset);
+    hitPaint..color = startColor;
     onCollision = (Set<Vector2> intersectionPoints, HitboxShape other) {
       if (other.component is ScreenCollidable) {
         hitPaint..color = startColor;
       } else {
-        hitPaint..color = hitColor;
+        hitPaint..color = hitColor.withOpacity(0.8);
```
Comment (erickzanardo) [inline examples/lib/stories/collision_detection/multiple_shapes.dart:177]:
```suggestion
        hitPaint.color = hitColor.withOpacity(0.8);
```

File: packages/flame/lib/src/geometry/collision_detection.dart [hunk 4]
```diff
@@ -52,8 +95,16 @@ Set<Vector2> intersections(
...
+        _shapeHashes.add(_combineHashCodes(shapeA, shapeB));
+      } else {
+        _handleShapeCollisionEnd(shapeA, shapeB);
       }
     }
   }
   return result;
 }
+
+// Note that this might result in hash collisions
```
Comment (erickzanardo) [inline packages/flame/lib/src/geometry/collision_detection.dart:107]: This is an interesting approach, butI imagine that this collision could result in some problems right? Or is this not something to worry about?

**Review by erickzanardo — COMMENTED** [review]
Some nit comments and one concern.

I think the docs needs to be updated to include the new callback right?

File: packages/flame/lib/src/geometry/collision_detection.dart (hunk 4 above)
Comment (spydon) [reply packages/flame/lib/src/geometry/collision_detection.dart]: Came up with a better idea.

**Review by erickzanardo — APPROVED** [review]
LGTM

File: examples/lib/stories/collision_detection/collision_detection.dart [hunk 5]
```diff
@@ -6,6 +6,19 @@ import 'circles.dart';
 import 'multiple_shapes.dart';
 import 'only_shapes.dart';
 
+const basicInfo = '''
+An example with many hitboxes that move around on the screen and during
+collisions they change color depending on what it is that they have collided
+with. 
+
+The snowman, the component built with three circles on top of each other, work
```
Comment (luanpotter) [inline examples/lib/stories/collision_detection/collision_detection.dart:14]:
```suggestion
The snowman, the component built with three circles on top of each other, works
```

**Review by luanpotter — APPROVED** [review]
LGTM

**Pull request merged**